### PR TITLE
[validate-modules] Allow argspec modifiers to be a part of docstring

### PR DESCRIPTION
**Pull request by NilashishC**
Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

##### SUMMARY
- In order to dynamically generate the module argspec from docstring we would require all keys that argspec supports to be a valid in docstring as well. 
- Currently, `validate-modules` complains if docstring contains any of the argspec modifiers since those are not a part of the defined schema. This patch attempts to add it there.

##### ISSUE TYPE

- Test Pull Request

##### COMPONENT NAME
validate_modules/schema.py

### CODE REVIEW

**Comment by abadger** [comment]
If I'm reading this right, this adds fields from the argspec like mutually_exclusive to each indivdual option and to each individual suboption.  Is that correct?

If so, I don't think this is quite right.  But I need a little more information to see what the solution is.  Do you have an example of a plugin which is using suboptions and modifiers where you can show how it currently works with arg_spec and how you translated it to the DOCUMENTATION strings?  I think that a before and after like that would help.

Some things that I think I see but want examples to verify:

* I think that the argument_spec_modifiers should also be added to the doc_schema.  This is because the toplevel options have their modifiers set at the level above the options (in code, in the `ArgumentSpecValidator`). 
* It feels like there should be a wrapper around the things listed in suboptions to hold the modifiers (just like the modifiers for options are held in the doc_schema, not in the options themselves) however, that's probably a large incompatibility with the current doc_schema and may not be how the current arg_spec code is structured....
  * A middle ground might be to allow argument_spec_modifiers for suboptions to be non-empty only if suboptions is non-empty?

Note that adding these fields needs to be coordinated with the ansible-core (to possibly update ansible-doc and the test code which you are changing inthis PR), the galaxy/AH team (since they want to produce docs for the plugins), and the docs team (code for generating the plugin docs lives in https://github.com/ansible-comunity/antsibull ). (@felixfontein and I are the main maintainers of antsibull)

**Comment by felixfontein** [comment]
CC @russoz since he's interested in this topic (https://github.com/ansible-collections/community.general/pull/2682, https://github.com/ansible-collections/community.general/pull/1992).

I would really prefer if there is an ansible-core supported way to do this, instead of every collection inventing their own way to obtain the argument spec from `DOCUMENTATION`.

And if these extra conditions can be specified in the documentation, there should be validation which ensures that they coincide with the conditions in the argument spec. (We should also think about requiring them to be documented in `DOCUMENTATION`.)

**Comment by russoz** [comment]
Thanks @felixfontein for the mention.

TL;DR: Having one single source of truth for argument specs is inherently a good coding/maintenance practice, and even though the complete consolidation would be something for a very long term plan, we can take steps into that direction today. 

The motivation for my interest is that we have two different sources of truth for (mostly) the same information, which causes many *synchronization* issues that have to be addressed by using sanity-checks, or by forcing developers to keep those two data-sets in sync. Cases where the module has a very long list of parameters (not uncommon, especially in cloud-provisioning modules), this makes adding a line in `ignore-X.Y.txt` a much more appealing option than translating one format to another. Or it might cause developers to abandon their ideas because of the bureaucracy/boilerplate task - which would be a worse scenario, because it impacts our community's ability to evolve.

That said, I have tried to somehow consolidate `DOCUMENTATION` and `argument_spec`. PR https://github.com/ansible-collections/community.general/pull/1992 was a first attempt to generate `argument_spec` out of the documentation, by using Ansible's internal code to parse docs. I wasn't by then aware of the `import` limitations in modules' code. In the second attempt (for which there is no PR), I tried to do the same from *within* `AnsibleModule` itself, only to find out that the same limitations apply to it. So, the third attempt (https://github.com/ansible-collections/community.general/pull/2682) does not import internal Ansible code, but rather it defers to the developer the responsibility of indicating the documentation blobs to be used and, for the time being, doing that only inside the `ModuleHelper` - a convenience wrapper around `AnsibleModule` by yours truly used by a only a handful of modules for the time being.

The part where that initiative and this one overlap is that the DOCS lack things like `no_log` and some deprecations attributes (*modifiers?* I haven't spent much time in the Ansible codebase itself, not sure of the terminology there). If we had all the attributes/modifiers of `argument_spec` available in the documentation, we could think of a long term plan to migrate all the argument specification to the documentation. This, of course, in the scope of new-style python modules. In time we could extend that to Powershell modules as well.

**Comment by NilashishC** [comment]
The Ansible Network team has been using the [`cli_rm_builder`](https://github.com/ansible-network/cli_rm_builder) tool to scaffold out boiler-plate code 
for new modules since early last year. As a part of the scaffolding process, the tool reads an externally supplied YAML docstring (or pulls the existing docstring from a module) and feeds it to a plugin that converts it into Python argspec.
This has been extremely beneficial for the module development process since the network modules often have very large and complex argspecs and as such, maintaining parity between module docstring and the argspec is challenging and potentially error-prone when done manually.

The [`to_argspec`](https://github.com/ansible-network/cli_rm_builder/blob/main/plugins/filter/to_argspec.py) filter does the docstring to argspec conversion. It internally uses the [`convert_doc_to_ansible_module_kwargs()`](https://github.com/ansible-collections/ansible.netcommon/blob/main/plugins/module_utils/network/common/utils.py#L780-L789) method. When updating the argspec for an existing module, the [`get_doc`](https://github.com/ansible-network/cli_rm_builder/blob/main/plugins/filter/get_doc.py) filter is used, which picks up the docstring from the module file and converts it into Python argspec with the aforementioned filter. We follow the practice of never writing or updating the argspecs manually, instead we make changes to the docstring (the "source of truth") and let the builder tool do its job.

As an example, we can refer to the [`nxos_bgp_global`](https://github.com/ansible-collections/cisco.nxos/blob/main/plugins/modules/nxos_bgp_global.py) module. We have been defining the argspec modifiers for top-level options while creating the module object itself. 

```
module = AnsibleModule(
        argument_spec=Bgp_globalArgs.argument_spec,
        mutually_exclusive=[["config", "running_config"]],
        required_if=[
            ["state", "merged", ["config"]],
            ["state", "replaced", ["config"]],
            ["state", "rendered", ["config"]],
            ["state", "parsed", ["running_config"]],
        ],
        supports_check_mode=True,
    )
```
Our long term goal is to eliminate the need of a separate argspec file and dynamically generate it while creating the module object.

The only issue for us right now is `validate-modules` complaining about argspec modifiers in docstring. I'm happy to discuss this further!

**Comment by abadger** [comment]
Where is your example of what nxos_bgp_global looks like after this
change?  Looking at how the present module is using arg spec modifiers, i
can only see a need for toplevel modifiers (which your change does not
add). Modifiers of suboptions are not used

**Comment by NilashishC** [comment]
@abadger Oh, I misunderstood. We don't have any code in the upstream network collections where a suboption has argspec modifiers in docstring because of the ansible-test restriction. But let me share an example from my fork:

Module docstring: https://github.com/NilashishC/nxos/blob/argspec_test/plugins/modules/nxos_route_maps.py#L303
Generated Argspec: https://github.com/NilashishC/nxos/blob/argspec_test/plugins/module_utils/network/nxos/argspec/route_maps/route_maps.py#L339-L340

Thanks!

**Comment by abadger** [comment]
@NilashishC Hey, sorry for taking a while to get back to this.  I looked at the examples you have from your fork.  I think that what I said here still stands: https://github.com/ansible/ansible/pull/74873#issuecomment-854107712  So things I can think of that need to be done:

- [ ] Add the argument_spec_modifiers to doc_schema as well (since toplevel options have these values set from there)
- [ ] Make the same changes to the schema in antsibull: https://github.com/ansible-community/antsibull/tree/main/antsibull/schemas/docs
- [ ] Propose how these should show up in documentation (Easiest way to review would probably be a PR that modifies the plugin template: https://github.com/ansible-community/antsibull/blob/main/antsibull/data/docsite/plugin.rst.j2 and then ask acozine and samccann to review the output of that [you may have to generate the output from your fork so they can see it])
- [ ] You may have to do the same for ansible-doc output but ansible-doc might just spit out the raw fields and let the user decide how to interpret them... A core team member would be better equipped to tell you precisely what ansible-doc does or does not need.

**Comment by abadger** [comment]
I think this is a start for the second item above: https://gist.github.com/abadger/3d47ae0ead289943758a07cc5555a99d

For it to be finished, someone needs to figure out how to express required_by  and there needs to be tests added to the functional test suite that show that it works at least for the happy path.

**Comment by samccann** [comment]
Adding to the list of https://github.com/ansible/ansible/pull/74873#issuecomment-880850893 -

- [ ] verify it works with galaxy-ng since that also is a heavy user of the DOCUMENTATION block.

**Comment by felixfontein** [comment]
I think the sanity tests should also make sure that when required_if etc. are documented in `DOCUMENTATION`, that they equal the versions in the argument spec.

**Comment by felixfontein** [comment]
Another aspect is whether these things should be allowed in plugin documentation as well, and if they're allowed there, whether they will be used for validation (since plugin docs == argument spec, as opposed to module docs).

**Comment by NilashishC** [comment]
The [last commit](https://github.com/ansible/ansible/pull/74873/commits/f2641e1c61c176f8621e2a047547dc0ddac0b113) adds support for rendering the top-level argspec modifiers in ansible-doc.

While looking into action item #1, I found out that validate-modules doesn't fail even if the module docstring has top-level keys that are undefined in doc schema. 

Both of these can be tested on my [fork](https://github.com/NilashishC/nxos/blob/argspec_test/plugins/modules/nxos_route_maps.py). 

@felixfontein I believe we are doing the same for the plugins in [ansible.utils](https://github.com/ansible-collections/ansible.utils) (i.e., deriving and validating argspec from docstring).

**Comment by felixfontein** [comment]
> While looking into action item #1, I found out that validate-modules doesn't fail even if the module docstring has top-level keys that are undefined in doc schema.

Yes, that's an unfortunate result of #73707. I created #74602 to fix that, but apparently as long as `attributes` isn't finalized, this isn't going to get merged...

> @felixfontein I believe we are doing the same for the plugins in [ansible.utils](https://github.com/ansible-collections/ansible.utils) (i.e., deriving and validating argspec from docstring).

I meant Ansible's [ConfigManager](https://github.com/ansible/ansible/blob/devel/lib/ansible/config/manager.py), which parses DOCUMENTATION and uses it to validate options for plugins.

**Comment by felixfontein** [comment]
> > While looking into action item #1, I found out that validate-modules doesn't fail even if the module docstring has top-level keys that are undefined in doc schema.
> 
> Yes, that's an unfortunate result of #73707. I created #74602 to fix that, but apparently as long as `attributes` isn't finalized, this isn't going to get merged... **Edit:** it looks like it is now finalized, so hopefully #74602 or something equivalent can get merged soon.

And by now it has been merged. Which means that from now on, these must be explicitly handled in your PR.

**Comment by samccann** [comment]
@NilashishC - what is the status of this work? Are you targetting this for core-2.12 (which has feature freeze in approx 2.5 weeks)

**Comment by jborean93** [comment]
Closing due to inactivity. If this is still something you wish to move forward with please open a new PR.
